### DISCUSSION TITLE
Implementing a Custom Repeated Sampler

### DISCUSSION THREAD

**Discussion by juhyeonkim95** [mitsuba-renderer/mitsuba3#682] [Q&A] [answered]
Hi,

I want to implement a new sampler in C++ code, but I'm having difficulties.

What I want to implement is a repeated sampler.

For example, if spp=8, next_1d() of independent sampler creates [r1, r2, r3, r4, r5, r6, r7, r8] where r1-r8 are independent random numbers in [0,1].

I want to implement a repeated sampler that creates [r1, r2, r3, r4, r1, r2, r3, r4] or  [r1, r1, r2, r2, r3, r3, r4, r4]  (both are okay).

I tried to use dr::repeat or dr::gather, but was unsuccessful.

Any help would be appreciated.

**Comment by njroussel** [comment]
Hi @juhyeonkim95 

Nothing immediately comes to mind as to why this wouldn't be feasible.
> I tried to use dr::repeat or dr::gather, but was unsuccessful.

This should have worked, what exactly did you try? The naive implementation would be to just wrap the `independent` sampler and every `next_1d()` would simply gather the first half from it with a repeating index.

**Reply by juhyeonkim95** [reply]
I tried something like this, but got an error message with `RuntimeError: jit_var_gather(): cannot gather from a placeholder variable!`

```
Float next_1d(Mask active = true){
        Assert(seeded());
        Float r = m_rng.template next_float<Float>(active);
        if constexpr(dr::is_array_v<Float>){
            UInt32 sample_indices = current_sample_index();
            r = dr::gather<Float>(r, sample_indices / 2);
        }
        return r;
}
```

**Reply by njroussel** [reply, marked as answer]
Ah, indeed. In short, we can't use `gather` inside of a recorded loop (basically the main integrator loop).

The solution would then be to construct the sampler to have the repetition.
The `inc` variables defines multiples RNS "streams" in the PCG32 sampler. It is initialized [here](https://github.com/mitsuba-renderer/drjit/blob/master/include/drjit/random.h#L59). You should be able to modify it such that is is repeating (it is a `public` member).

**Reply by juhyeonkim95** [reply]
It worked well! Thank you so much.
